library/core/src/num/dec2flt/float.rs

```diff
@@ -0,0 +1,203 @@
+    fn from_u64(v: u64) -> Self;
+
+    /// Performs a raw transmutation from an integer.
+    fn from_u64_bits(v: u64) -> Self;
+
+    /// Get a small power-of-ten for fast-path multiplication.
+    fn pow10_fast_path(exponent: usize) -> Self;
+
+    /// Returns the category that this number falls into.
+    fn classify(self) -> FpCategory;
+
+    /// Returns the mantissa, exponent and sign as integers.
+    fn integer_decode(self) -> (u64, i16, i8);
+}
+
+impl RawFloat for f32 {
+    const INFINITY: Self = f32::INFINITY;
+    const NEG_INFINITY: Self = f32::NEG_INFINITY;
+    const NAN: Self = f32::NAN;
+    const NEG_NAN: Self = -f32::NAN;
+
+    const MANTISSA_EXPLICIT_BITS: usize = 23;
+    const MIN_EXPONENT_ROUND_TO_EVEN: i32 = -17;
+    const MAX_EXPONENT_ROUND_TO_EVEN: i32 = 10;
+    const MIN_EXPONENT_FAST_PATH: i64 = -10; // assuming FLT_EVAL_METHOD = 0
+    const MAX_EXPONENT_FAST_PATH: i64 = 10;
+    const MAX_EXPONENT_DISGUISED_FAST_PATH: i64 = 17;
+    const MINIMUM_EXPONENT: i32 = -127;
+    const INFINITE_POWER: i32 = 0xFF;
+    const SIGN_INDEX: usize = 31;
+    const SMALLEST_POWER_OF_TEN: i32 = -65;
+    const LARGEST_POWER_OF_TEN: i32 = 38;
+
+    fn from_u64(v: u64) -> Self {
+        v as _
+    }
+
+    fn from_u64_bits(v: u64) -> Self {
+        f32::from_bits((v & 0xFFFFFFFF) as u32)
+    }
+
+    fn pow10_fast_path(exponent: usize) -> Self {
+        #[allow(clippy::use_self)]
+        const TABLE: [f32; 16] =
+            [1e0, 1e1, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8, 1e9, 1e10, 0., 0., 0., 0., 0.];
+        TABLE[exponent & 15]
+    }
+
+    /// Returns the mantissa, exponent and sign as integers.
+    fn integer_decode(self) -> (u64, i16, i8) {
+        let bits = self.to_bits();
+        let sign: i8 = if bits >> 31 == 0 { 1 } else { -1 };
+        let mut exponent: i16 = ((bits >> 23) & 0xff) as i16;
+        let mantissa =
+            if exponent == 0 { (bits & 0x7fffff) << 1 } else { (bits & 0x7fffff) | 0x800000 };
+        // Exponent bias + mantissa shift
+        exponent -= 127 + 23;
+        (mantissa as u64, exponent, sign)
+    }
+
+    fn classify(self) -> FpCategory {
+        self.classify()
+    }
+}
+
+impl RawFloat for f64 {
+    const INFINITY: Self = f64::INFINITY;
+    const NEG_INFINITY: Self = f64::NEG_INFINITY;
+    const NAN: Self = f64::NAN;
+    const NEG_NAN: Self = -f64::NAN;
+
+    const MANTISSA_EXPLICIT_BITS: usize = 52;
+    const MIN_EXPONENT_ROUND_TO_EVEN: i32 = -4;
+    const MAX_EXPONENT_ROUND_TO_EVEN: i32 = 23;
+    const MIN_EXPONENT_FAST_PATH: i64 = -22; // assuming FLT_EVAL_METHOD = 0
+    const MAX_EXPONENT_FAST_PATH: i64 = 22;
+    const MAX_EXPONENT_DISGUISED_FAST_PATH: i64 = 37;
+    const MINIMUM_EXPONENT: i32 = -1023;
+    const INFINITE_POWER: i32 = 0x7FF;
+    const SIGN_INDEX: usize = 63;
+    const SMALLEST_POWER_OF_TEN: i32 = -342;
+    const LARGEST_POWER_OF_TEN: i32 = 308;
+
+    fn from_u64(v: u64) -> Self {
+        v as _
```

Comment:
Is it ok that this can lose precision?

Reply:
Yep. It's only called when the number is <= `F::MAX_MANTISSA_FAST_PATH`, which ensures that there is no loss in precision here. It's only used for the fast-path algorithm. I'll add a comment documenting this.